### PR TITLE
feat(locale): add time zone adaptation according to tenant user setting

### LINKED ISSUE #1856: 国际化 - 时区
- 检查测试所有输出输入的时间

### PR DESCRIPTION
**Pull request by Qi-Cui**
### Description

Fixes #1856

### Checklist

- [ ] 填写 PR 描述及相关 issue (write PR description and related issue)
- [ ] 代码风格检查通过 (code style check passed)
- [ ] PR 中包含单元测试 (include unit test)
- [ ] 单元测试通过 (unit test passed)
- [ ] 本地开发联调环境验证通过 (local development environment verification passed)

### CODE REVIEW

File: src/bk-user/bkuser/apps/timezone/set_timezone.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from bkuser.apps.tenant.models import TenantUser
+
+
+@receiver(user_logged_in)
+def set_timezone(sender, request, user, **kwargs):
+    """每次登录读取并添加用户时区信息"""
+
+    # 如果当前会话没有时区信息，则尝试从租户用户模型中获取
+    if not request.session[settings.TIME_ZONE_SESSION_KEY]:
+        tenantuser = TenantUser.objects.filter(id=user.username).first()
```
Comment (narasux) [inline src/bk-user/bkuser/apps/timezone/set_timezone.py:25]: Nit: tenant_user

File: src/bk-user/bkuser/apps/timezone/apps.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+"""
+
+from django.apps import AppConfig
+
+
+class BkI18nAppConfig(AppConfig):
+    name = "bkuser.apps.timezone"
+
+    def ready(self):
+        import bkuser.apps.timezone.set_timezone  # noqa
```
Comment (narasux) [inline src/bk-user/bkuser/apps/timezone/apps.py:19]: django 风格是把接收信号的函数放在 handlers.py 里面

```python
def ready(self):
    # Install signal handlers
    from . import handlers  # noqa: F401
```

Comment (Qi-Cui) [reply src/bk-user/bkuser/apps/timezone/apps.py]: done

File: src/bk-user/bkuser/apps/timezone/set_timezone.py (hunk 1 above)
Comment (Qi-Cui) [reply src/bk-user/bkuser/apps/timezone/set_timezone.py]: done

File: src/bk-user/bkuser/apps/timezone/handlers.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    # 如果当前会话没有时区信息，则尝试从租户用户模型中获取
+    if not request.session[settings.TIME_ZONE_SESSION_KEY]:
+        tenant_user = TenantUser.objects.filter(id=user.username).first()
+        time_zone = settings.TIME_ZONE
+
+        if tenant_user:
+            time_zone = tenant_user.time_zone
+
+        request.session[settings.TIME_ZONE_SESSION_KEY] = time_zone
```
Comment (nannan00) [inline src/bk-user/bkuser/apps/timezone/handlers.py:31]: 引入 signals 有点复杂了，建议看看是否有中间件就能解决的方式？

File: src/bk-user/bkuser/apps/timezone/middleware.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+from django.conf import settings
+from django.utils import timezone
+
+
+class TimezoneMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        tzname = request.session.get(settings.TIME_ZONE_SESSION_KEY)
```
Comment (nannan00) [inline src/bk-user/bkuser/apps/timezone/middleware.py:22]: tzname = request.user.get_property("time_zone")

Comment (nannan00) [reply src/bk-user/bkuser/apps/timezone/middleware.py:22]: bool(request.user and request.user.is_authenticated)

File: src/bk-user/bkuser/apps/timezone/handlers.py (hunk 3 above)
Comment (Qi-Cui) [reply src/bk-user/bkuser/apps/timezone/handlers.py]: done，现在用中间件处理

File: src/bk-user/bkuser/common/middlewares.py [hunk 5]
```diff
@@ -29,3 +33,17 @@ def __call__(self, request):
...
+class TimeZoneMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        # 如果时区非空，则激活当前时区。如果时区为空，则使用默认时区
+        if request.user and request.user.is_authenticated and (tzname := request.user.get_property("time_zone")):
+            timezone.activate(pytz.timezone(tzname))
+        else:
+            timezone.deactivate()
```
Comment (nannan00) [inline src/bk-user/bkuser/common/middlewares.py:47]: 考虑：timezone.deactivate() 执行后 Django 是如何处理时间时区的？会有什么影响

Comment (Qi-Cui) [reply src/bk-user/bkuser/common/middlewares.py:47]: 屏幕截图 2024-08-09 095919
deactivate 后会读取 settings.TIME_ZONE 里定义的默认值

**Review by nannan00 — APPROVED** [review]

**Pull request merged**